proposals/pattern-variables.md

```diff
@@ -0,0 +1,158 @@
+# Variable declarations under disjunctive patterns
+
+## Summary  
+
+## Motivation
+
+## Specification
+
+### Scope
+
+#### Pattern variable redeclaration
+
+Introducing a new concept similar to *name hiding*, but these names can possibly reference either of variables based on the result of the pattern-matching at runtime.
+
+There's two degrees of enforcement as to whether variables may be redeclared:
+
+- Pattern variables within pattern boundaries *must* be redeclared because assignment of such variables depend on the order of evaluation which is undefined for patterns.
+- Pattern variables outside pattern boundaries *can* be redeclared, definitive assignment rules for the containing expression specify where such variables are usable.
+
+For a *disjunctive_pattern* of the form `left_pattern or right_pattern`:
+- Pattern variables declared in the *left_pattern* **must be redeclared** in the *right_pattern*; permitting:
+
+  ```cs
+  e is (0, var x) or (var x, 0)
+  ```
+
+For a *logical_or_expression* of the form `left_expr || right_expr`:
+- Pattern variables declared in *left_expr* **can be redeclared** in the *right_expr*; permitting:
+
+  ```cs
+  e is (0, var x) || e is (var x, 0)
+  ```
+  Note: This is the expression variation of the previous case.
+
+For a *negated_pattern* of the form `not pattern_operand`:
+- Pattern variables declared in the *pattern_operand* **must be redeclared** inside any other *negated_pattern* in the containing pattern; permitting:
```

Comment:
"... inside any other *negated_pattern*..."

What if there's more than one nested negated_pattern? `e is not (0, var x) and not not (0, 0)`

Reply:
Haven't thought this through yet. To keep the proposal simple, we can add this at some later point if there's enough support to allow such patterns.